wrappers/services.go

```diff
@@ -471,116 +471,180 @@ type EnsureSnapServicesOptions struct {
-	// note, sysd is not used when preseeding
-	sysd := systemd.New(systemd.SystemMode, inter)
-
-	if opts == nil {
-		opts = &EnsureSnapServicesOptions{}
-	}
-
-	// we only consider the global EnsureSnapServicesOptions to decide if we
-	// are preseeding or not to reduce confusion about which set of options
-	// determines whether we are preseeding or not during the ensure operation
-	preseeding := opts.Preseeding
-
-	// modifiedUnitsPreviousState is the set of units that were modified and the previous
+// ensureSnapServicesContext is the context for EnsureSnapServices.
+// EnsureSnapServices supports transactional update/write of systemd service
+// files and slice files. A part of this is to support rollback of files and
+// also keep track of whether a restart of systemd daemon is required.
+type ensureSnapServicesContext struct {
+	// snaps, observeChange, opts and inter are the arguments
+	// taken by EnsureSnapServices. They are here to allow sub-functions
+	// to easier access these, and keep parameter lists shorter.
+	snaps         map[*snap.Info]*SnapServiceOptions
+	observeChange ObserveChangeCallback
+	opts          *EnsureSnapServicesOptions
+	inter         Interacter
+
+	// note: is not used when preseeding is set in opts.Preseeding
+	sysd systemd.Systemd
+	// modifiedSystemServices/modifiedUserServices keeps track of whether a systemd
+	// daemon reload is necessary, either for the system level and/or user level.
+	modifiedSystemServices bool
+	modifiedUserServices   bool
+	// modifiedUnits is the set of units that were modified and the previous
 	// state of the unit before modification that we can roll back to if there
 	// are any issues.
 	// note that the rollback is best effort, if we are rebooted in the middle,
 	// there is no guarantee about the state of files, some may have been
 	// updated and some may have been rolled back, higher level tasks/changes
 	// should have do/undo handlers to properly handle the case where this
 	// function is interrupted midway
-	modifiedUnitsPreviousState := make(map[string]*osutil.MemoryFileState)
-	var modifiedSystem, modifiedUser bool
+	modifiedUnits map[string]*osutil.MemoryFileState
+}
 
-	defer func() {
-		if err == nil {
-			return
-		}
-		for file, state := range modifiedUnitsPreviousState {
-			if state == nil {
-				// we don't have anything to rollback to, so just remove the
-				// file
-				if e := os.Remove(file); e != nil {
-					inter.Notify(fmt.Sprintf("while trying to remove %s due to previous failure: %v", file, e))
-				}
-			} else {
-				// rollback the file to the previous state
-				if e := osutil.EnsureFileState(file, state); e != nil {
-					inter.Notify(fmt.Sprintf("while trying to rollback %s due to previous failure: %v", file, e))
-				}
+// restore is a helper function which should be called in case any errors happen
+// during the write/update of systemd files
+func (es ensureSnapServicesContext) restore() {
```

Comment:
I would use a pointer here:
```suggestion
func (es *ensureSnapServicesContext) restore() {
```
Otherwise a copy of the structure is created when you call this method (https://stackoverflow.com/questions/25382073/defining-golang-struct-function-using-pointer-or-not)

Reply:
Ah awesome! Thanks for teaching me something new :-)